File: sorting/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        for i in range(len(nums)):
            
            if i>0 and nums[i-1] == nums[i]:
                continue
            nums1 = nums[i]
            l,r = i+1, len(nums)-1
            while l < r:
                total = nums1 + nums[l] + nums[r]
                if total >0:
                    r-=1
                if total <0:
                    l+=1
                if total == 0:
                    res.append([nums1, nums[l], nums[r]])
                    l+=1
                while l<r and nums[l-1] == nums[l]:
                    l+=1
        return res

File: array/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums); // Sort the array to use two-pointer technique

        for (int i = 0; i < nums.length - 2; i++) {
            // Skip duplicate elements
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            int curr = nums[i];
            int l = i + 1;
            int r = nums.length - 1;

            while (l < r) {
                int curr_sum = curr + nums[l] + nums[r];

                if (curr_sum == 0) {
                    res.add(Arrays.asList(curr, nums[l], nums[r]));
                    l++;
                    r--;

                    // Skip duplicate elements
                    while (l < r && nums[l] == nums[l - 1]) {
                        l++;
                    }
                    while (l < r && nums[r] == nums[r + 1]) {
                        r--;
                    }
                } else if (curr_sum < 0) {
                    l++;
                } else {
                    r--;
                }
            }
        }

        return res;
    }
}